WIP - build(docs-infra): update docs examples to latest v9 and Ivy

gkalpak opened the pull request:
Jira issue: [FW-1609](https://angular-team.atlassian.net/browse/FW-1609)

##
Blocked on:
- [x] ~~#33484 (+updating `aio/tools/examples/shared/boilerplate/elements/package.json` according to ` aio/tools/examples/shared/boilerplate/cli/package.json`)~~ _Landed in #36015._
- [x] ~~#34508~~ (+updating `zone.js` dependency version) _Landed in #36015._
- [x] ~~#35778~~
- [x] ~~#35780~~
- [x] ~~#35858~~
- [x] ~~#36015~~ (orignally part of this PR - split off for easier review/merging)
- [x] ~~#36143~~ (orignally part of this PR - split off for easier review/merging)
- [x] ~~#36483 (orignally part of this PR - split off for easier review/merging)~~

##
Fixes #34382, #34383 and #35289.

This PR may also partially address: #35776

mary-poppins commented:
You can preview c6f5e72 at https://pr34374-c6f5e72.ngbuilds.io/.

mary-poppins commented:
You can preview d3acac0 at https://pr34374-d3acac0.ngbuilds.io/.

mary-poppins commented:
You can preview 8b02ec3 at https://pr34374-8b02ec3.ngbuilds.io/.

mary-poppins commented:
You can preview 94dced4 at https://pr34374-94dced4.ngbuilds.io/.

mary-poppins commented:
You can preview bdb74cb at https://pr34374-bdb74cb.ngbuilds.io/.

mary-poppins commented:
You can preview 1987ff7 at https://pr34374-1987ff7.ngbuilds.io/.

mary-poppins commented:
You can preview 9f7cbc2 at https://pr34374-9f7cbc2.ngbuilds.io/.

mary-poppins commented:
You can preview be630c0 at https://pr34374-be630c0.ngbuilds.io/.

mary-poppins commented:
You can preview 7294b9f at https://pr34374-7294b9f.ngbuilds.io/.

mary-poppins commented:
You can preview 334bb0b at https://pr34374-334bb0b.ngbuilds.io/.

mary-poppins commented:
You can preview 92a917a at https://pr34374-92a917a.ngbuilds.io/.

mary-poppins commented:
You can preview 8255f1a at https://pr34374-8255f1a.ngbuilds.io/.

IgorMinar reviewed (changes requested):
as discussed in https://github.com/angular/angular/pull/35091#discussion_r373775396 let's also switch the npm fw packages to the dist/packagest-dist VE distribution

mary-poppins commented:
You can preview a4893e2 at https://pr34374-a4893e2.ngbuilds.io/.

mary-poppins commented:
You can preview 04bc693 at https://pr34374-04bc693.ngbuilds.io/.

mary-poppins commented:
You can preview 475f1c9 at https://pr34374-475f1c9.ngbuilds.io/.

mary-poppins commented:
You can preview b625c18 at https://pr34374-b625c18.ngbuilds.io/.

mary-poppins commented:
You can preview d4fab26 at https://pr34374-d4fab26.ngbuilds.io/.

mary-poppins commented:
You can preview d052718 at https://pr34374-d052718.ngbuilds.io/.

mary-poppins commented:
You can preview 49b1ee1 at https://pr34374-49b1ee1.ngbuilds.io/.

mary-poppins commented:
You can preview 963eac0 at https://pr34374-963eac0.ngbuilds.io/.

mary-poppins commented:
You can preview b221938 at https://pr34374-b221938.ngbuilds.io/.

mary-poppins commented:
You can preview 04f904a at https://pr34374-04f904a.ngbuilds.io/.

mary-poppins commented:
You can preview 73cc101 at https://pr34374-73cc101.ngbuilds.io/.

mary-poppins commented:
You can preview bcb53ca at https://pr34374-bcb53ca.ngbuilds.io/.

gkalpak commented:
Marking this as blocked on #35778 and #35780, since they need to land first.

mary-poppins commented:
You can preview df58abd at https://pr34374-df58abd.ngbuilds.io/.

mary-poppins commented:
You can preview b475e53 at https://pr34374-b475e53.ngbuilds.io/.

mary-poppins commented:
You can preview 4d9a91c at https://pr34374-4d9a91c.ngbuilds.io/.

mary-poppins commented:
You can preview 73b5f11 at https://pr34374-73b5f11.ngbuilds.io/.

mary-poppins commented:
You can preview f8b5fb2 at https://pr34374-f8b5fb2.ngbuilds.io/.

mary-poppins commented:
You can preview cd97096 at https://pr34374-cd97096.ngbuilds.io/.

mary-poppins commented:
You can preview e1e56a1 at https://pr34374-e1e56a1.ngbuilds.io/.

mary-poppins commented:
You can preview 24dc855 at https://pr34374-24dc855.ngbuilds.io/.

mary-poppins commented:
You can preview e5d6683 at https://pr34374-e5d6683.ngbuilds.io/.

mary-poppins commented:
You can preview ed915a3 at https://pr34374-ed915a3.ngbuilds.io/.

mary-poppins commented:
You can preview 6eac4cc at https://pr34374-6eac4cc.ngbuilds.io/.

mary-poppins commented:
You can preview 89a37a0 at https://pr34374-89a37a0.ngbuilds.io/.

mary-poppins commented:
You can preview c095256 at https://pr34374-c095256.ngbuilds.io/.

mary-poppins commented:
You can preview dbae03a at https://pr34374-dbae03a.ngbuilds.io/.

mary-poppins commented:
You can preview d8b3206 at https://pr34374-d8b3206.ngbuilds.io/.

gkalpak committed:
build(docs-infra): add missing `build` npm script for `universal` docs example ZIP archive

Previously, the `package.json` template used in the ZIP archive of the
`universal` example that we offer for download missed the `build` npm
script.

This commit updates the template for the `universal` docs example to
include the `build` npm script.

NOTE:
The `build` npm script is already included in
`aio/tools/examples/shared/boilerplate/universal/package.json`, but it
was removed by the example zipper.

gkalpak committed:
style(docs-infra): clean up Tour-of-Heroes examples

I noticed these minor styling issues while aligning the `universal`
examples with the `toh-pt6` example (in a subsequent commit).

gkalpak committed:
refactor(docs-infra): update `main.ts` in Tour-of-Heroes examples to match latest CLI

Update the `main.ts` files in Tour-of-Heroes examples to match what
would be generated by the latest CLI.

gkalpak committed:
fix(docs-infra): align `universal` example with `toh-pt6`

As mentioned in the `universal` guide, the `toh-pt6` examples is the
starting poitn for the `universal` example. However, the two examples
had become out-of-sync, because some fixes/changes were made to the
Tour-of-Heroes examples.

This commit ports these changes to the `universal` example.

gkalpak committed:
refactor(docs-infra): update `universal` example to match latest CLI

Update the `universal` example (and related files) to match what would
be generated by the latest CLI.

gkalpak committed:
build(docs-infra): update `@types/express-serve-static-core` to avoid error in `universal` example

Previously, building the `universal` example failed with:
```
node_modules/@types/express/index.d.ts(90,50): error TS2694: Namespace '".../@types/express-serve-static-core/index"' has no exported member 'Params'.
node_modules/@types/express/index.d.ts(90,64): error TS2694: Namespace '".../@types/express-serve-static-core/index"' has no exported member 'ParamsDictionary'.
```

This commit fixes the error by upgrading
`@types/express-serve-static-core` to a newer version.
See DefinitelyTyped/DefinitelyTyped#40905 for more details.

gkalpak committed:
test(docs-infra): add tests for `universal` docs example

Previously, there were no tests for the `universal` docs example. This
meant that the project was not tested at all (not even ensuring that it
can be built successfully).

This commit adds e2e tests for the `universal` example (ported from
`toh-pt6` and cleaned up) and also verifies that the project can be
built successfully (including the server).

gkalpak committed:
docs: minor fixes/improvements in the `universal` guide

gkalpak committed:
docs: remove unneeded code from `universal` example/guide

In the past, server-side rendered apps needed to convert URLs used in
API requests to absolute when rendering on the server. Originally, this
was handled in the `universal` guide and corresponding example app by
modifying the `HeroService` to use `APP_BASE_HREF` to derive the
absolute URL.

In #28956, the guide was updated to show an improved method: Specifying
an `HttpInterceptor` that took care of converting the URLs to absolute.
That interceptor was only provided when rendering the app on the server.
By mistake, the corresponding example app was not updated along with the
guide.

Since `@nguniversal/*` v7.1.0, it is no longer necessary to convert the
URLs to absolute inside the app. This is handled in the `@nguniversal`
libs (see angular/universal#897).

This commit updates the example app to emove unnecessary code and
modifies the guide to mention the issue with absolute URLs, but explain
that developers only need to worry about it when not using one of the
`@nguniversal/*-engine` packages.

gkalpak committed:
WIP - add support for per example-type ViewEngine boilerplate

gkalpak committed:
refactor(docs-infra): update `i18n` example to match latest CLI

Update the `i18n` example (and related files) to match what would be
generated by the latest CLI.

gkalpak committed:
TEMP

mary-poppins commented:
You can preview c8ce569 at https://pr34374-c8ce569.ngbuilds.io/.

CaerusKaru commented on aio/content/guide/universal.md:
```diff
@@ -334,3 +266,37 @@ The following Node.js Express code routes all remaining requests to `/dist`, and
+
+If, for some reason, you are not using an `@nguniversal/*-engine` package, you may need to handle it yourself.
+
+One solution would be to provide the full URL to your application on the server (e.g. as a server-side only extra provider) and write an interceptor that can retrieve this value and prepend it to the request URL.
```
As much as is this is probably the most approachable solution, I'd like to recommend providing the `url` option to `renderModule` from `@angular/platform-server`, since that initializes the URL platform-wide.

gkalpak replied:
Nice! I didn't know about that; thx for bringing it up :+1:
I have pushed a fixup commit to recommend this approach instead: https://github.com/angular/angular/pull/34374/commits/3072e2e2f3109036020f650a78905802da1bc5e7

gkalpak committed:
build(docs-infra): switch docs examples to Ivy

The docs examples are switched to Ivy. On CI, the tests are run with
both Ivy and ViewEngine.

Partially addresses:
[FW-1609](https://angular-team.atlassian.net/browse/FW-1609)

gkalpak committed:
test(docs-infra): fix unit tests and run them for specific docs examples on CI

Previously, only e2e tests were run for docs examples on CI. As a
result, unit tests (which are included in the zipped archives we provide
for users to download and play with the examples locally) were often
outdated and broken.

This commit configures specific docs examples that have meaningful unit
tests to run them on CI (via the `run-example-e2e.js` script). Where
necessary, the unit tests are fixed to ensure they pass and reflect the
changes in the corresponding component/service.
This commit also removes some auto-generated unit tests that are not
meaningful (e.g. make trivial assertions, such that a component instance
is truthy) and are often broken anyway (e.g. because the corresponding
component has been changed in ways that make the tests fail).

gkalpak committed:
build(docs-infra): renamed `e2e` property of `example-config.json` to `tests`

Each docs example has an `example-config.json` configuration file. Among
other things, this file can be used to specify what commands to run in
order to test the example. (If not specified, the `run-example-e2e.js`
script will run a default `yarn e2e` command.)

Previously, the property specifying the test commands was called `e2e`.
This is because in the past only e2e tests were run for docs examples.
Since recently, some examples may specify commands for other types of
tests (such as unit tests). Therefore, calling the property that holds
the list of test commands `e2e` no longer makes sense and can be
misleading to people looking at the configuration files.

This commit renamed the property to the more generic `tests`. In the
future, the `run-example-e2e.js` script (and corresponding npm script)
should be renamed and refactored to also avoid giving the impression
that only e2e tests are run.

Discussed in:
https://github.com/angular/angular/pull/36143#discussion_r395148379

gkalpak committed:
build(docs-infra): update docs examples `package.json` templates wrt `core-js`

The `core-js` dependency is no longer included in `package.json` for
`cli`-type examples, but only for the `systemjs` ones. This commit
updates the `package.json` templates to reflect that (and also updates
the `npm-packages` guide accordingly).

gkalpak committed:
fixup! docs: remove unneeded code from `universal` example/guide

gkalpak commented:
The final bit was addressed in #36924, so this can be closed.
